Mathlib/NumberTheory/FLT/Three.lean

```diff
@@ -38,3 +42,135 @@ theorem fermatLastTheoremThree_case_1 {a b c : ℕ} (hdvd : ¬ 3 ∣ a * b * c)
   rcases cube_of_not_dvd hdvd.1.2 with hb | hb <;>
   rcases cube_of_not_dvd hdvd.2 with hc | hc <;>
   rw [ha, hb, hc] <;> decide
+
+/-- To prove `FermatLastTheoremFor 3`, we may assume that `3 ∣ c`. -/
+theorem fermatLastTheoremThree_of_three_dvd_c
+    (H : ∀ a b c : ℤ, a ≠ 0 → b ≠ 0 → c ≠ 0 → 3 ∣ c → a ^ 3 + b ^ 3 ≠ c ^ 3) :
+    FermatLastTheoremFor 3 := by
+  intro a b c ha hb hc habc
+  by_cases h1 : 3 ∣ a * b * c
+  swap
+  · exact fermatLastTheoremThree_case_1 h1 habc
+  rw [Nat.Prime.dvd_mul (Nat.prime_three), Nat.Prime.dvd_mul (Nat.prime_three)] at h1
+  rcases h1 with ((⟨k, hk⟩ | ⟨k, hk⟩) | ⟨k, hk⟩)
+  · refine H (-(c : ℤ)) b (-(a : ℤ)) (by simp [hc]) (by simp [hb]) (by simp [ha]) ?_ ?_
+    · exact ⟨-k, by simp [hk]⟩
+    · rw [Odd.neg_pow (by decide), Odd.neg_pow (by decide), add_comm, ← sub_eq_add_neg,
+        sub_eq_iff_eq_add, add_comm, ← sub_eq_add_neg, eq_sub_iff_add_eq, add_comm]
+      exact_mod_cast habc
+  · refine H a (-(c : ℤ)) ((-(b : ℤ))) (by simp [ha]) (by simp [hc]) (by simp [hb])
+      ⟨-k, by simp [hk]⟩ ?_
+    rw [Odd.neg_pow (by decide), Odd.neg_pow (by decide), ← sub_eq_add_neg, sub_eq_iff_eq_add,
+        add_comm, ← sub_eq_add_neg, eq_sub_iff_add_eq]
+    exact_mod_cast habc
+  · exact H a b c (by simp [ha]) (by simp [hb]) (by simp [hc]) ⟨k, by simp [hk]⟩
+      (by exact_mod_cast habc)
+
+variable {a b c : ℕ} (ha : 3 ∣ a) (hb : 3 ∣ b) (hc : 3 ∣ c) (hF : a ^ 3 + b ^ 3 = c ^ 3)
+
+lemma three_dvd_gcd_of_dvd_a_of_dvd_c : 3 ∣ ({a, b, c} : Finset ℕ).gcd id := by
+  have hb : 3 ∣ b := by
+    have : 3 ∣ (b : ℤ) ^ 3 := by
+      replace hF : (a : ℤ) ^ 3 + (b : ℤ) ^ 3 = (c : ℤ) ^ 3 := by exact_mod_cast hF
+      rw [add_comm, ← eq_sub_iff_add_eq] at hF
+      rw [hF]
+      exact dvd_sub (dvd_pow (by exact_mod_cast hc) (by decide))
+        (dvd_pow (by exact_mod_cast ha) (by decide))
+    exact Int.coe_nat_dvd.1 <| Int.prime_three.dvd_of_dvd_pow this
+  refine Finset.dvd_gcd (fun x hx ↦ ?_)
+  simp only [Finset.mem_insert, Finset.mem_singleton] at hx
+  rcases hx with (hx | hx | hx)
+  · exact hx ▸ ha
+  · exact hx ▸ hb
+  · exact hx ▸ hc
+
+lemma three_dvd_gcd_of_dvd_a_of_dvd_b : 3 ∣ ({a, b, c} : Finset ℕ).gcd id := by
```

Comment:
There is a lot of repetition here with these three lemmas that all basically say the same thing.
It might help to work with `a^3 + b^3 + c^3 = 0` instead of `a^3 + b^3 = c^3`, to make the symmetry more apparent. Then the two further lemmas can be more easily reduced to the first.